enhancements/gep/001-loadbalancer/README.md

```diff
@@ -0,0 +1,119 @@
+
+## Motivation
+At the moment, Gravity is using CoreDNS to determine which master to send requests to,
+but DNS is not a good tool for balancing traffic.
+This proposal aims to provide the following improvements:
+  1. Evenly distribute the load between all apiserver replicas
+  1. Improve the stability and scalability of the cluster
+  1. Nodes will react faster to master failures
+  1. Cluster upgrades will be faster and more efficient
+
+### Goals
+When creating a cluster, the user can choose the load balancer to be either external or internal.
+External loadbalancer is not controlled or configured by Gravity, it is the user's responsibility.
+The internal loadbalancer, on the contrary, is controlled and configured by Gravity, the planet-agent is responsible for this.
+When using an internal loadbalancer, a loadbalancer (HAProxy) will be installed for each node in the cluster including the master nodes.
+When adding/removing master nodes, the agent on each node should automatically reflect this in the local loadbalancer configuration.
+All Kubernetes components (kubelet, scheduler, control-manager, kube-proxy) will use the loadbalancer to access the apiserver.
+
+### Non-Goals
+
+## Proposal
+Change DNS balancer to a software loadbalancer (HAProxy).
+CoreDNS will resolve `leader.telekube.local` to either the external address (domain name or IP) of the loadbalancer (subject to which configuration?) or to `127.0.0.1` when using the internal loadbalancer.
+Docker image references should continue pointing to internal registry at `leader.telekube.local:5000`.
+To achieve this, docker registry will use port 5001 with port 5000 left to the loadbalancer.
+
+### User Stories
+#### Story 1
+As a cluster administrator, I can configure the loadbalancer when creating a cluster. If I set the type of loadbalancer to external, I can provide the external address in configuration.
+#### Story 2
+As a cluster administrator, I can change the loadbalancer settings on a running cluster.
+
+### Risks and Mitigations
+As the `leader.telekube.local` domain will not point to a master node, the applications using this address should be updated accordingly.
+
+## Design Details
+An additional section will be added to the cluster manifest for setting up the loadbalancer:
+```yaml
+kind: Cluster
+apiVersion: cluster.gravitational.io/v2
+loadbalancer:
+  # default value is internal
+  type: "external|internal"
+  # gravity uses this field when type is external
+  externalAddress: "IP address | DNS address"
+```
+The following section will be added to the cluster configuration:
+```yaml
+kind: ClusterConfiguration
+version: v1
+spec:
+  loadbalancer:
+    # default value is internal
+    type: "external|internal"
+    # gravity uses this field when type is external
+    externalAddress: "IP address | DNS address"
+```
+The docker image references should not be changed, and they should start with `leader.telekube.local:5000/`. 
```

Comment:
One of the problems gravity has, is it uses gravity-site to synchronize the registries. So on a multi-master install, I believe it only bootstraps the first registry, and then waits for gravity-site to sync the container images once it's loaded. On the current design, this causes problems if there is a failover of the leader within the first minute of install, as the registry may be empty. If this happens before gravity-site starts, then it can't start and sync the registry.